cmd/mt-backfill/main.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	l "log"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/grafana/globalconf"
+	"github.com/grafana/metrictank/api"
+	"github.com/grafana/metrictank/cluster"
+	"github.com/grafana/metrictank/idx"
+	"github.com/grafana/metrictank/idx/memory"
+	"github.com/grafana/metrictank/input"
+	inKafka "github.com/grafana/metrictank/input/kafkamdm"
+	"github.com/grafana/metrictank/logger"
+	"github.com/grafana/metrictank/mdata"
+	"github.com/grafana/metrictank/mdata/cache"
+	cassandraStore "github.com/grafana/metrictank/store/cassandra"
+	"github.com/raintank/dur"
+	"github.com/raintank/schema"
+	"github.com/raintank/schema/msg"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// metrictank
+	aggMetrics  *mdata.AggMetrics
+	metricIndex idx.MetricIndex
+	apiServer   *api.Server
+	inputKafka  input.Plugin
+	store       mdata.Store
+
+	// config file
+	confFile = flag.String("config", "/etc/metrictank/metrictank.ini", "config file path")
+
+	// Data: the following configs are the same with the normal metrictank configs:
+	chunkMaxStaleStr  = flag.String("chunk-max-stale", "1m", "chunk max stale age.")
+	metricMaxStaleStr = flag.String("metric-max-stale", "5m", "metric max stale age.")
+	gcIntervalStr     = flag.String("gc-interval", "2m", "gc interval.")
+	publicOrg         = flag.Int("public-org", 0, "org Id")
+
+	// backfilling
+	backfillEnd = "backfillEnd" // message used by input stream to indicate the end of backfilling
+	lastRcvTime int64           // epoch time when the previous kafka message was received
+	mux         sync.Mutex      // mutex to protect lastRcvTime
+	handler     inputHandler    // input message handler to track the last kafka receive event and handles kafka messages
+)
+
+// a kafka mesage handler that implements the input.Handler interface
+type inputHandler struct {
+	dHandler input.DefaultHandler // default handler that processes metric metadata and points
+	finished chan bool            // this will be set if a message contains backfillEnd
+}
+
+func newInputHandler(metrics mdata.Metrics, metricIndex idx.MetricIndex, pluginName string) inputHandler {
+	dh := input.NewDefaultHandler(metrics, metricIndex, pluginName)
+	return inputHandler{
+		dHandler: dh,
+		finished: make(chan bool),
+	}
+}
+
+// input.Handler interface
+func (ih inputHandler) ProcessMetricData(metric *schema.MetricData, partition int32) {
+	if strings.Contains(metric.Name, backfillEnd) {
```

Comment:
Hmm, I see why this is here, but it feels kinda hacky. Could we instead require that the kafka topic is filled beforehand and this tool just goes until it hits the end? Or reduce the "no data" timeout below and use that?

Reply:
Yes this seems hacky, i can remove this.

You mean we should increase the "no data" timeout below(the current timeout is 3 min)? The tool will simply wait until there is no more data from kafka.

Reply:
Yeah, like 10 seconds without data seems like plenty (but also maybe make it configurable)

Reply:
sure, I will make this configurable